rig/place_and_route/place/sa/python_kernel.py

```diff
@@ -0,0 +1,432 @@
+        self.machine = machine
+        self.random = random
+
+        self.has_wrap_around_links = self.machine.has_wrap_around_links()
+
+        # Location-to-Vertices: A lookup {(x, y): [vertex, ...], ...} giving
+        # the set of vertices on a given chip.  Chips which are not in the
+        # machine are excluded from this lookup.
+        self.l2v = {xy: [] for xy in self.machine}
+        for vertex, location in iteritems(self.placements):
+            self.l2v[location].append(vertex)
+
+        # Vertices-to-Nets: A lookup {vertex: [Net, ...], ...}, gives a list of
+        # nets of which the given vertex is a member.
+        self.v2n = {v: [] for v in self.vertices_resources}
+        for net in nets:
+            for v in net:
+                if net not in self.v2n[v]:
+                    self.v2n[v].append(net)
+
+    def run_steps(self, num_steps, distance_limit, temperature):
+        num_accepted = 0
+        deltas = []
+        for _ in range(num_steps):
+            swapped, delta = _step(self.movable_vertices,
+                                   distance_limit, temperature,
+                                   self.placements, self.l2v, self.v2n,
+                                   self.vertices_resources,
+                                   self.fixed_vertices,
+                                   self.machine,
+                                   self.has_wrap_around_links,
+                                   self.random)
+            num_accepted += 1 if swapped else 0
+            deltas.append(delta)
+        mean = sum(deltas) / float(len(deltas))
+        std = math.sqrt(sum((v-mean)**2 for v in deltas) / len(deltas))
+
+        cost = sum((_net_cost(net, self.placements, self.has_wrap_around_links,
+                              self.machine)
+                    for net in self.nets), 0.0)
+
+        return num_accepted, cost, std
+
+    def get_placements(self):
+        return self.placements
+
+
+def _net_cost(net, placements, has_wrap_around_links, machine):
+    """Get the cost of a given net.
+
+    This function, in principle at least, should estimate the total network
+    resources consumed by the given net. In practice this estimate is based on
+    the size of the bounding-box of the net (i.e. HPWL). This should be
+    improved at some later time to better account for the effects of large
+    fan-outs.
+
+    Parameters
+    ----------
+    net : :py:class:`rig.netlist.Net`
+    placements : {vertex: (x, y), ...}
+    has_wrap_around_links : bool
+    machine : :py:class:`rig.place_and_route.Machine`
+
+    Returns
+    -------
+    float
+    """
+    # This function is by far the hottest code in the entire algorithm, as a
+    # result, small performance improvements in here can have significant
+    # impact on the runtime of the overall algorithm. As an unfortunate side
+    # effect, this code is rather ugly since many higher-level constructs (e.g.
+    # min/max) are outrageously slow.
```

Comment:
Out of interest did you look at this wondrous document? https://wiki.python.org/moin/PythonSpeed/PerformanceTips?highlight=%28%28PythonSpeed|PerformanceTips%29%29

(Warning: benchmarks were run on a Pentium I)!